### ISSUE TITLE
ENH: `optimize.root_scalar(..., method='newton')` could get root in 1/3 fewer objective calls when numeric differentiation is used

### ISSUE THREAD

**Issue by nickodell** [scipy/scipy#21165]
### Is your feature request related to a problem? Please describe.

Something I noticed while debugging #21148 is that some points are evaluated twice. The Newton method needs two things to find the next point: the value of the function at the current point, and the derivative. When the derivative is being evaluated with numeric differentiation, the Newton method makes three calls to the objective function: f(x), f(x), f(x + h). The second two are done in order to get the derivative at the current point, but the first and second calls could be shared.

Example reproducer:

```
from scipy import optimize
import traceback
import sys

def f_p_pp(x):
    print(x)
    return x**3 - 1

sol = optimize.root_scalar(f_p_pp, x0=10)
```

Output:

```
10
[10]
[10.00000015]
6.670000050811767
[6.67000005]
[6.67000015]
4.454159241760294
[4.45415924]
[4.45415931]
2.986240987351696
[2.98624099]
[2.98624103]
2.0282064581930923
[2.02820646]
[2.02820649]
1.4331692485858107
[1.43316925]
[1.43316927]
1.1177331892286098
[1.11773319]
[1.11773321]
1.011965675147062
[1.01196568]
[1.01196569]
1.0001409269281298
[1.00014093]
[1.00014094]
1.0000000198582968
[1.00000002]
[1.00000003]
1.0000000000000007
[1.]
[1.00000001]
```

### Describe the solution you'd like.

The `approx_derivative()` function accepts an f0 argument, to allow a method which already knows the function value to provide it to `approx_derivative()` so that it is not re-evaluated.

The proposed solution would be that `optimize.root_scalar()` adds an optional f0 keyword argument to `fprime()`, so that solvers which know f0 can provide it.

There are seven other solvers for `optimize.root_scalar()`:

* bisect
* brentq
* brenth
* ridder
* toms748
* secant
* halley

It would be good to look into these too and see if the same optimization can be done.

### Describe alternatives you've considered.

None.

**Comment by fancidev** [comment]
Nice observation. For bracketed solvers (like Brent) I sometimes run into a situation where the end points of the initial bracket is already evaluated before calling the solver, but they are evaluated again by the solver upon entry. The extra evaluation could be saved.

If this is to be handled systematically, would it be a potential solution for the user function (or a wrapper of it) to cache the last (few) evaluated values? But then there is the question of how many recent values to cache.

**Comment by swstkm** [comment]
Hi, I'm new here. I'm not sure about the bracketed solvers, but I'd like to improve the others. Would that be useful, i.e. simply adding an optional f0 argument and using it if available?

**Comment by nickodell** [comment]
> Hi, I'm new here. I'm not sure about the bracketed solvers, but I'd like to improve the others.

Yes, that would be useful. As for the bracketed solvers, I investigated this today, and none of them do this a significant amount. Fixing newton would fix 95% of the problem.

>Would that be useful, i.e. simply adding an optional f0 argument and using it if available?

After thinking about it for a bit, I favor @fancidev's suggestion more:

>If this is to be handled systematically, would it be a potential solution for the user function (or a wrapper of it) to cache the last (few) evaluated values? But then there is the question of how many recent values to cache.

To be more specific, you only need to cache the last value. This covers the vast majority of cases.

Please let me know if you have any questions, or need any help getting started.

**Comment by swstkm** [comment]
Okay, I'll work on the Newton first and then attempt the rest.

**Comment by andyfaff** [comment]
There is already `optimize._differentiable_functions.ScalarFunction` that may be appropriate, it has a caching behaviour. Would save reinventing the wheel.
